RFC: Vulnerability response runbook (#20)
bytecodealliance/rfcs: accepted/vulnerability-response-runbook.md

# Bytecode Alliance Vulnerability Response Runbook

> 

Reviewer: This runbook is a great baseline for situations where a vulnerability has been reported or discovered privately. Should we add some brief guidance here on what a reader should do if an unpatched vulnerability is under active exploitation? Or if details of a private (or previously unknown) vulnerability were made public without coordination.

It could be something simple like this...
```suggestion
This document describes a series of steps to be followed by the Bytecode Alliance team when responding to a vulnerability that is discovered or reported privately.

This document does not cover cases where vulnerabilities are publicly disclosed without coordination. Or cases where unpatched vulnerabilities are under active exploitation. In those cases, contact security@bytecodealliance.org.
```
Or we could build out more process for these cases proactively.

## Select an Incident Manager

> Pick an individual who will track that the the entire process in this document isfollowed.

Reviewer: ```suggestion
Pick an individual who will track that the entire process in this document is followed.
```

By default, the person inside Bytecode Alliance who discovers or learns of the
vulnerability first has this duty, until they explicitly hand if off to
someone else.

## Identifying the Vulnerability

Figure out what bytecode alliance repository it lives in.

The vulnerability probably lives in one single repository in the Bytecode
Alliance ecosystem. If fixing it requires coordinated work across two

> repositories, this guide may not work.

Reviewer: What should the reader do then? Maybe email security@bytecodealliance.org or ping folks on Zulip?

> repositories, this guide may not work.

Author: ```suggestion
repositories, this guide may not work. In situations not covered by this guide, please
email security@bytecodealliance.org to seek guidance.
```

requesting a CVE.

### Request a CVE

Once you have a draft description, you should have GitHub request a CVE on
your behalf. During US west coast business hours, this typically takes under 2
hours to process. If it takes more than a day, ping Bytecode Alliance partners
who may have contacts at GitHub, where they can inquire on why the process is
stuck. You can keep working on everything while this request is pending.

## Invite collaborators to the Security Advisory Draft

The incident manager is responsible for selecting trustworthy individuals to collaborate
on authoring the security advisory and its patch.  The set of collaborators should be
kept as small as possible. It should include an individual who is a code reviewer on
each Bytecode Alliance project affected by the vulnerability.

GitHub Security Advisories let you invite individuals to collaborate on the draft.
Collaborators will be able to view and edit the advisory, and push and pull commits
to the private repository where the patch is under development.

## Collaborate on a patch

Use the security release PR features in the GitHub Security Advisory to

> collaboate on a patch. GitHub will create a private fork of the repository on

Reviewer: ```suggestion
collaborate on a patch. GitHub will create a private fork of the repository on
```

your behalf and invite the advisory collaboators to it. 

In the private fork repository, create a single PR which will track the patch

> to resolve the vulnerabilty. This patch is automatically merged into the

Reviewer: ```suggestion
to resolve the vulnerability. This patch is automatically merged into the
```

public repository's default branch when the disclosure is made public.

Cloud-based CI (github actions or otherwise) will not run on the security
advisory fork. Collaborators are responsible for running CI locally before
pushing to the PR branch.

Use the PR review features to indicate approval for the patch.

### Releases as part of the patch

Wasmtime has committed that security issues will be applied to the current
version of Wasmtime, and released as patch releases: see [Wasmtime 1.0
RFC](https://github.com/bytecodealliance/rfcs/blob/main/accepted/wasmtime-one-dot-oh.md#backports---security-fixes)

Other projects without an existing policy should do the same, unless they have
a very good reason for issuing backports.

The patch PR should increment any version numbers required to make the patch
release. The merged PR should be immediately publishable to package registries
when the disclosure is made public. Create a merge commit and run your package
publishing tools in dry-run mode to verify this ahead-of-time as much as
possible.

## Advanced disclosure to stakeholders

The incident manager is responsible for identifying the downstream making advanced

> disclosure to trustworthy stakeholders. All stakeholders who recieve advanced disclosure

Reviewer: ```suggestion
disclosure to trustworthy stakeholders. All stakeholders who receive advanced disclosure
```

must maintain secrecy of the vulnerability until public disclosure is made.

The incident manager determines who is a trustworthy stakeholder in a vulnerability,
and what sort of advanced disclosure to make to each stakeholder. This could mean
giving a stakeholder complete access to the advisory draft and patch, or providing only
more limited information, such as a warning that a vulnerability exists and a public
disclosure is expected on a given date.

Bytecode Alliance members are encouraged to socialize which projects they rely on
in production so that incident managers know to consider them for advanced disclosure.

> No individual or orginization is entitled to advanced disclosure. Advanced disclosure is

Reviewer 2: ```suggestion
No individual or organization is entitled to advanced disclosure. Advanced disclosure is
```

> not guaranteed to orginizations by membership in the Bytecode Alliance, nor is it exclusive

Reviewer 2: ```suggestion
not guaranteed to organizations by membership in the Bytecode Alliance, nor is it exclusive
```

to Bytecode Alliance members. Selecting stakeholders for advanced disclosure appropriately

> is critical for maintaining trust in the Bytecode Alliance and its projects.

Reviewer 3: The cautiousness in this section is very close to "security through obscurity", as it tries to limit shared knowledge by asserting that this augments trust within the Alliance. The counter argument is that not trusting by default (and only include arbitrarily trusted people within the Alliance) sounds more like the opposite of trust.

Additionally, basing the selection of contacts on socialization has limits: it doesn't look like it would scale well with an increasing number of organizations in the Bytecode Alliance, times an increasing number of projects hosted by the Bytecode Alliance. In fact, it's unclear what should be the topics of socialization, and with which depth one should share implementation details that might also require an individual organization's discretion (think NDA).

I think there's an opportunity to do better here. We could imagine sharing light details with technical contacts of each Bytecode Alliance organization, on an internal but private secured channel (e.g. Signal, or any other equally valid broadcasting secured mechanism), that would make it possible for members to know if they're effectively affected or not. That could include, for example: patterns to look for in error logs, in the case of Wasmtime examples of `Config()` that would be affected by the security issue, metrics to watch and for which outliers could indicate potential exploitation in the wild. Other useful information could be shared according to the context around the security issue. After sharing that information, technical contacts could confirm whether they're impacted or not, and if impacted some extra actions could take place to ensure the issue might be mitigated for them (temporary patches for mitigations? getting early access to the CVE? request filters if Web/cloud providers? Each issue is likely to have its own set of temporary and useful countermeasures).

Of course this process doesn't pretend to be perfect and could be adapted. I think that in general, this would spur trust within the Bytecode Alliance and across its different members, by trusting by default instead of assuming the worst (notably that insiders would leak the exploit). Some entities might and will run Bytecode Alliance projects in production and they could be badly impacted if they had to schedule updates at the last minute, would they never hear about a security issue, because it was incorrectly decided (due to ineffective/incomplete/poor "socialization") that they would not be impacted.

Author: I hear your concerns, but I don't think I am the right person to address them. When I wrote this up, I my goal was to document the principles we used to navigate this in the past, rather than come up with an improved system. I will have to defer to @tschneidereit and other Bytecode Alliance leaders on this topic.

Reviewer 4: This RFC has been open and stalled for a while now, and I think it has a lot of really good stuff in it that is helpful for the people dealing with these rare security bugs. No, this RFC does not have all the answers and there are improvements we can make. However, it is still a super helpful resource for when folks are caught up in the midst of a security bug. And there is no reason we can't further refine our processes with future RFCs.

With all that said, would you be okay if we moved to merge this RFC, @bnjbvr?

Reviewer 3: Agreed that there's lots of goodness in this runbook!

I would be fine with merging, *if we removed those three paragraphs I commented about* (since there's been no discussion about the core issue I'm flagging here, and it looks quite important to me to, at the very least, discuss). Approving the RFC as it is right now would mean turning this into the default, widely accepted measure, and since communication has stalled as you point, anybody opening a PR to change that would likely face a similar stall situation, adding way more friction to changing the wording here. Does that sound reasonable?

Reviewer 4: That seems like a reasonable path forward to me. @pchickey sound good to you?

Author: Agreed, and done!

## Commit to a public disclosure date

Once you have confidence in a patch, and stakeholders have been pre-notified, it is
time to determine a date for public disclosure. Incident managers may discuss the public

> disclosure date with stakeholders who have recieved advanced to understand their feasibility

Reviewer: ```suggestion
disclosure date with stakeholders who have received advanced to understand their feasibility
```

and timelines for mitigation.

By default, public disclosure should be no more than 30 days after initial discovery
of the vulnerability. Incident managers are encouraged to select the shortest timeline
possible.

However, the incident manager may make an exception to the 30 day limit if the
nature of the vulnerability requires it. Reasons for exceptions may include the
technical feasibility of mitigating the vulnerability, or coordination with other
security disclosure processes. Do not make exceptions lightly: inappropriate exceptions
may undermine trust in the Bytecode Alliance and its projects.

## Notify sec-announce 

Once you have committed to a disclosure date, send an email to the
[`sec-announce@bytecodealliance.org` Google
Group](https://groups.google.com/a/bytecodealliance.org/g/sec-announce).

This email should use this template:

```pre
The Bytecode Alliance would like to announce the forthcoming security release
of the <project>.

The release will be made available on <date> at approximately <time> UTC.
Additionally, an advisory will be made available on <date> at approximately
<time> at https://github.com/advisories.

> The highes severity issue fixed in this release is <severity>, based on the

Reviewer: ```suggestion
The highest severity issue fixed in this release is <severity>, based on the
```

classification scheme defined in the OpenSSL Security Policy.

<non-affected projects> Bytecode Alliance projects are unaffected.
```

## Finalize the security advisory

Improve the draft description of the security advisory into a public facing
document.

> Security Advisories may not assign blame to individuals or orginizations

Reviewer: ```suggestion
Security Advisories may not assign blame to individuals or organizations
```

who introduced or contributed to the vulnerability.

Security Advisories may credit individuals who contributed to discovering and
remediating the vulnerability. Use the credit feature of the GitHub Security
Advisory to do so. This feature requires individuals to consent to being

> credited through their GitHub account.

Reviewer 2: If the individual who discovered the vulnerability doesn't have a github account what is the intended procedure? Provide their name and or website in the security advisory body? Or something else?

> credited through their GitHub account.

Author: ```suggestion
credited through their GitHub account. The incident manager can decide how to accommodate creditees
who do not wish to use the GitHub mechanism.
```
